### ISSUE TITLE
Uploaded file datetime

### ISSUE THREAD

**Issue by yar229** [yar229/WebDavMailRuCloud#111]

**Comment by yar229** [comment]
Везде разброд и шатание.
* Far+NetBox отсылает кривую дату, https://github.com/michaellukashov/Far-NetBox/issues/261
* TotalCommander + webdav плагин вообще не заморачивается установкой даты файла.
* CarotDAV вместо стандартного свойства `getlastmodified` использует MS'ое `Win32CreationTime`

**Comment by VictorVG** [comment]
Плагин отсылает тот таймстам который дала локальная ось.

**Comment by VictorVG** [comment]
НетБокс осылает тот таймстам который выдала ваша локальная ОС, посему вопросы к ней.:)

**Comment by yar229** [comment]
done [#111_upload_datetime](https://github.com/yar229/WebDavMailRuCloud/tree/%23111_upload_datetime)

**Comment by yar229** [comment]
Заметки по поводу:
на данный момент для файла из облака есть только одна дата/время, судя по имени возвращаемых данных из запроса - это время модификации файла.
При закачке файла Far/NetBox и CarotDav отправляют только дату/время создания файла,
TotalCommander и BitKinex при закачке не отправляют запрос на модификацию времени создания/модификации файла, и оно всегда будет равно времени на момент закачки файла.

Соответственно, для файла в облаке в GUI обеих программ будет видно время создания файла, а на локальных дисках (с настройками по умолчанию) видно не время создания файла, а время его модификации.

**Comment by VictorVG** [comment]
Я тут у себя ставил простой опыт с батником (была идея использовать время):

```
@echo off
setlocal
call :time %1
echo Time for %1 is %retn%.
exit
:time
set retn=%~dpt1
goto:EOF
```
и ОС, точнее говорить командный процессор, возвращал именно время последней записи исследуемого объекта, что по идее кажется разумным так как мы видим момент его изменения с одной стороны, а с другой это не всегда верно так как ФС могла изменить служебные атрибуты например произошло обращение по чтению, ОС зафиксировала его время в атрибуте, а после т.к.  ФС рассматривает файл и его атрибуты как единое логическое целое уже она сообщает нам "Файл изменился" и мы это видим на примере NTFS либо к примеру JP Software TakeCommand.epub - на сервере этот файл с SHA-256 `665b42f3b1a2613d3d2980b83b09a9528562d4c02a9de338d00325edc2913ff9 *TakeCommand.epub
 `появился 13.10.2017, но 15.11.2017 скрипт

```
@echo off
setlocal
if "%~1" =="" (goto mhlp)
if "%~1" == "-h" (goto mhlp)
if not "%~2" == "" (set dnl=%2) else (set dnl=c)
if exist JPSoft (move /y JPSoft JPSoft.old > nul)
if /i "%dnl%" == "W" (goto mwget)
:mcurl
for %%a in (
       tcc-rt.exe
       tcc.exe
       tcmd.exe
       tcmdupdate.aiu
       tcc-update.aiu
       tcc-rt-update.aiu
       TakeCommand.epub
       TakeCommand.pdf
       ) do curl -Rs --create-dirs --output ./JPSoft/%%a https://jpsoft.com/downloads/v%1/%%a
for %%a in (
       4autotray
       4console
       4utils
       ev
       onig32dll
       sysutils
       ) do curl -C - -RLs --create-dirs --output ./JPSoft/Plugins/%%a.zip ftp://vefatica.net/4plugins/%%a.zip
for %%a in (
       4autotray64
       4console64
       4utils64
       ev64
       onig64dll
       sysutils64
       ) do curl -C - -RLs --create-dirs --output ./JPSoft/Plugins/%%a.zip ftp://vefatica.net/4plugins/x64/%%a.zip
for %%a in (
       chill
       consearch
       htags
       htmldump
       iso8601
       popinfo
       popupfix
       qcal
       safechars
       say
       ) do curl -C - -RLs --create-dirs --output ./JPSoft/Plugins/%%a.zip http://prospero.unm.edu/dl/%%a.zip
for %%a in (
       event
       everything
       fedutils9
       ) do curl -C - -RLs --create-dirs --output ./JPSoft/Plugins/%%a.zip https://jpsoft.com/downloads/plugins/%%a.zip
curl -C - -RLs --create-dirs --output ./JPSoft/sdk/sdk.zip https://jpsoft.com/downloads/sdk/sdk.zip
goto mchk
:mwget
cd /d %windir%
if exist wget.exe (set wg=wget & goto mnext)
if exist wget64.exe (set wg=wget64 & goto mnext) else (goto mcurl)
:mnext
if exist curl-ca-bundle.crt (set prm=%wg%--ca-certificate=curl-ca-bundle.crt --no-hsts --ca-directory=./) else (set prm=%wg%--no-check-certificate --no-hsts)
for %%a in (
       tcc-rt.exe
       tcc.exe
       tcmd.exe
       tcmdupdate.aiu
       tcc-update.aiu
       tcc-rt-update.aiu
       TakeCommand.epub
       TakeCommand.pdf
       ) do %prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft https://jpsoft.com/downloads/v%1/%%a
for %%a in (
       4autotray
       4console
       4utils
       ev
       onig32dll
       sysutils
       ) do %prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft/Plugins ftp://vefatica.net/4plugins/%%a.zip
for %%a in (
       4autotray64
       4console64
       4utils64
       ev64
       onig64dll
       sysutils64
       ) do %prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft/Plugins ftp://vefatica.net/4plugins/x64/%%a.zip
for %%a in (
       chill
       consearch
       htags
       htmldump
       iso8601
       popinfo
       popupfix
       qcal
       safechars
       say
       ) do  %prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft/Plugins http://prospero.unm.edu/dl/%%a.zip
for %%a in (
       event
       everything
       fedutils9
       ) do %prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft/Plugins https://jpsoft.com/downloads/plugins/%%a.zip
%prm% -c -q -nH --keep-badhash -t5 -P %~dp0JPSoft/SDK https://jpsoft.com/downloads/sdk/sdk.zip
endlocal
cd /d %~dp0
:mchk
for %%a in (
       tcc-rt
       tcc
       tcmd
       ) do if not exist JPSoft\%%a.exe echo %%a setup not downloaded >> JPSoft\download.log
for %%a in (
       tcmdupdate
       tcc-update
       tcc-rt-update
       ) do if not exist JPSoft\%%a.aiu echo %% update script not downloaded >> JPSoft\download.log
for %%a in (
       epub
       pdf
       ) do if not exist JPSoft\TakeCommand.%%a echo Take Command %%a manual not downloaded >> JPSoft\download.log
for %%a in (
       4autotray
       4autotray64
       4console
       4console64
       4utils
       4utils64
       chill
       consearch
       ev
       ev64
       event
       everything
       fedutils9
       htags
       htmldump
       iso8601
       onig32dll
       onig64dll
       popinfo
       popupfix
       qcal
       safechars
       say
       sysutils
       sysutils64
       ) do if not exist JPSoft\Plugins\%%a.zip echo Plugin %%a not downloaded >> JPSoft\download.log
if not exist JPSoft\SDK\sdk.zip echo Plugin SDK not download >> JPSoft\download.log
if exist JPSoft\download.log (type JPSoft\download.log & pause)
exit
:mhlp
echo.
echo jpsoft-upd v2.1 script help:
echo.
echo SYNOPSIS:
echo.
echo jpsoft-upd PARAM1 PARAM2
echo.
echo download v21 use cURL:   jpsoft-upd 21 C
echo download v21 use Wget:   jpsoft-upd 21 w
echo display  help        : jpsoft-upd -h
echo display  help        : jpsoft-upd
echo.
echo Description:
echo.
echo Download latest JP Software TCC, TCMD, TCMD-RT, plugins and documentations.
echo.
echo History:
echo.
echo v2.1, 11.11.2017 19:32:02 +0300 - refactoring and some plug-ins added.
echo.
echo Command line switches:
echo.
echo PARAM1 is "Version" major, example 21 , or "-h" or run w/o parametes for help.
echo PARAM2 is "C" for use cURL v7.56 or never, "W" - use Wget v1.19 or never.
echo.
pause
exit /b 
```
скачает его нам с новой датой 15.11.2017 и только по сравнению SHA-256 мы увидим что изменений в файле нет. Это не единичный случай, а потому я для себя предпочту потратить время на сравнение SHA-256 или SHA-512 (этот алгоритм быстрее, но хэш естественно длиннее), но не стану ориентироваться по потенциально ненадёжному показателю ВРЕМЯ.

**Comment by yar229** [comment]
Показатель ненадёжный, но его используют для синхронизации каталогов.
Бохсним, что зачастую вместо modified зачастую посылается creation time, что поделать, лишь бы хоть что-то было.
Хэш файла, безусловно, надежнее, но для того, чтобы его посчитать, надо скачать файл.

**Comment by VictorVG** [comment]
Согласен.  Значит стоит придумать как обойти его ненадёжность. Доступ к логу сервера для исключения по нему операций типа "перекинули в другой каталог, передумали, вернули" у нас нет, да и просто создание симлинка или хардлинка на мгновенно изменит время модификации объекта за счёт того, что на стороне сервера над ним произвели файловую операцию. 

А если использовать не полное, а частичное сравнение в ряде точек тела файла? Например начало, середина, конец? Для мелких файлов из чтение для сверки при современных скоростях каналов не заметно, а вот на крупных так можно сэкономить трафик, да и вероятность того, что получаемый/пересылаемый файл совпадёт с существующим удастся снизить, а значит сократится число бесполезных операций и снизится износ оборудования.

**Comment by yar229** [comment]
Вообще, всегда есть хэш файла, который лежит в облаке и можно посчитать хэш локального файла. Это прекрасно в случае, если имеется контроль над клиентом.
В данном случае его получить ну вот никак - клиент говорит "PUT, вот тебе поток, положи его по такому-то пути в облаке", а потом - "PROPPATCH, вот дата файла"

**Comment by VictorVG** [comment]
Нет, удалённый  хэщ доступен клиенту не всегда, да и формат  в котором он там хранится может оказаться неприемлем для нас. Плюс воздействие помех на канал связи и оборудование - я не раз встречался со случаями когда помеха на входе триггера Шмидта (входной шинный формирователь, как раз он должен сводить влияние помех на канале связи к нулю) оказывалась достаточной для его срабатывания, хотя по идее  он в принципе не должен был на неё среагировать. И сиди думай сам ли неверно схему рассчитал или конструктора где напутали? А лишний бит исказит принятые нами данные, и если искажены контрольные данные то достоверность проверки стремится к нулю.

Потому я бы предпочёл ввести в цепочку "лишнее" звено счёта, чем неопределённость влияния помех на канале связи и по цепям питания аппаратуры. Особенно учитывая модный среди дилетантов "разгон" - результат работы схемы и момент её отказа непредсказуем, но повод для хвастовства перед приятелями есть, и своё ЭГО отлично тешится.:) Да, при разработке мы закладываем запас надёжности схемы на изменение условий среды и старение компонентов, только разгон его за часы сводит к нулю. Да и время между посылкой нами на сервер запроса "пришли хэш" и его получением (опять же при условии админ на сервер не запретил клиентам работать с хэшами под предлогом что это служебные данные или как "Я, Знаменитый Дмитрий Постригань, Автор Знаменитой Утилиты MHDD готов вернуть вам утраченные после аварии моих серверов данные после получения от вас оплаты за них!" (я это письмо получил лично от сего персонажа когда он грохнул свои сервера и стал требовать с людей денег за возвращение их данных)) может оказаться несоизмеримым со временем "лишнего" просчёта. Тут стоит смоделировать события во времени. Что расчёт матмодели событий покажет? Я когда то считал протокол CSMA/CD на задержки - в этом смысле он далеко не лучший, а мой вариант слямзил один из преподов кафедры и продал крупной фирме предложившей купить мои разработки за $5. Понятно я отказался, но ошибочный вариант (ошибку я устранил через год и она была принципиальной делавшей протокол неработоспособным) фирма пыталась использовать, но он так и не нашёл применения.

**Comment by yar229** [comment]
blocked by #116

**Comment by VictorVG** [comment]
В принципе можем найти обход, только каковы будут накладные затраты на его работу?

**Comment by yar229** [comment]
пока не думаю об этом, надо запилить #116

**Comment by VictorVG** [comment]
Вопросов нет. У самого сдача работы на носу а с генеральным спорить бесполезно.:)
